Default jdbc-ping cluster setup for distributed caches fails in Oracle

ryanemerson opened the pull request:
Closes #40784
Closes #41105

In general we rely upon Hibernate/Liquibase and the DB drivers to handle Schema case sensitivity for us, for example: https://github.com/keycloak/keycloak/blob/87fbe2327047bc8b29cee7e500c917c4846828f5/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/database/jpa/QuarkusJpaConnectionProviderFactory.java#L112

However, when creating the SQL query for JDBC_PING and `JpaRealmProvider#preRemove` we are explicitly quoting the name of the default schema making it case-sensitive.

When querying the default schema via Hibernate, the schema name is returned case insensitively in lower-case. Therefore, in the case of DBs that don't default to lower-case, e.g. Oracle, applying quotations around the name in the SQL will lead to a mismatch and the schema will not be found.

ryanemerson committed:
Default jdbc-ping cluster setup for distributed caches fails in Oracle

* Add DatabaseConfig to TestDatabase so the underlying DB can be
  configured per test
* Allow DB initScripts to be configured by tests

Closes #40784
Closes #41105

Signed-off-by: Ryan Emerson <remerson@redhat.com>

ahus1 reviewed (changes requested):
The old code made it case sensitive by adding the quote all the time, and now it is never adding them so it will break for schemata that contain special characters like `-`.

Can you figure how hibernate handles it that it adds the quotes conditionally?  

We don't test schemata well today, so it might be good to add one.

ryanemerson commented:
Good points @ahus1, my  initial attempt was overly simplistic. I have reworked `JpaUtils#getTableNameForNativeQuery` so that instead of querying hibernate for the configured default schema, we rely upon `JpaConnectionProviderFactory#getSchema` and apply quotes to the schema only when `GlobalConfiguration.PRESERVE_SCHEMA_CASE` has been set to true.

I have also created #41105 to also track the issue with `JpaRealmProvider#preRemove` so that any users who run into this will be able to see there's a fix.

For testing I have introduced two new tests:

- `CustomSchemaTest` - Which ensures that #41105 is resolved for all tested DBs
- `JdbcPingCustomSchemaTest` - Which ensures that JDBC_PING is correctly configured with Postgres and a cluster forms as expected. I have also manually executed this with `KC_TEST_DATABSE=oracle` to make sure it works as expected.

In order to be able to create a Postgres schema before Keycloak startup, I have added the ability for initScripts to be executed by the Database. If/when you're happy with the proposed changes, we should ask the testing team to review this part of the work.

ahus1 commented on model/jpa/src/main/java/org/keycloak/connections/jpa/util/JpaUtils.java:
```diff
@@ -56,9 +58,10 @@ public class JpaUtils {
-        String schema = (String) em.getEntityManagerFactory().getProperties().get(HIBERNATE_DEFAULT_SCHEMA);
         final Dialect dialect = em.getEntityManagerFactory().unwrap(SessionFactoryImpl.class).getJdbcServices().getDialect();
-        return (schema==null) ? tableName : dialect.openQuote() + schema + dialect.closeQuote() + "." + tableName;
+        IdentifierHelper identifierHelper = em.getEntityManagerFactory().unwrap(SessionFactoryImpl.class).getJdbcServices().getJdbcEnvironment().getIdentifierHelper();
+        String schema = em.getEntityManagerFactory().unwrap(SessionFactoryImpl.class).getSessionFactoryOptions().getDefaultSchema();
+        return (schema==null) ? tableName : identifierHelper.toIdentifier(schema).render(dialect) + "." + tableName;
```
The initial suggestion used the Liquibase setting, but I think this is mixing JPA and Liquibase items up more than they should be mixed up. And when quoting the schema, we should apply the same logic as Hibernate applies - only quote it when there are special characters. 

I suggest to keep the rest of the code untouched. I added a separate commit, but then I found that there were some conflicts in the intermediate commit, so I squashed your and my changes. 

Please have a look and let me know.

ryanemerson replied:
Thanks for pushing the change. I agree that your approach is cleaner and it makes sense to utilise JPA/Hibernate here. I wasn't aware of `IdentifierHelper`, good to know :+1:

ahus1 reviewed (commented):
I would like to keep the original logic mostly unchanged and not mix JPA logic with Liquibase options. See below for a suggestion. 

Please review your test case if it still makes sense ... maybe we should should force a schema name that enforces quotes? Feel free to join a 1:1 on discussing the tests.

ryanemerson committed:
Refactor existing test and add additional test-case for schemas containing hyphen

Signed-off-by: Ryan Emerson <remerson@redhat.com>

ryanemerson committed:
Ensure that org.keycloak.tests.db.* tests are executed first so subsequent CI tests don't fail due to a differrent DB name being configured

Signed-off-by: Ryan Emerson <remerson@redhat.com>

ryanemerson commented:
The use of `KC_TEST_DATABASE_REUSE` is causing problems with subsequent steps in the store tests as we now have multiple DBs being kept alive and the old tests are sometimes connecting to the DB provisioned with a different schema name etc. I'm working on a fix, but I'll change the PR back to draft until this is resolved.

ryanemerson committed:
Ignore KC_TEST_DATABASE_REUSE when Lifecycle.Global is not configured

Signed-off-by: Ryan Emerson <remerson@redhat.com>

ryanemerson committed:
Wait for Ryuk container to terminate before attempting to retrieve database port

Signed-off-by: Ryan Emerson <remerson@redhat.com>

ahus1 committed:
Review

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 reviewed (approved):
Thank you for this change!

vaceksimon commented:
Thanks for letting me know @ahus1 I would like to do a little cleanup to unify the database config style with the rest of the framework. See #41319
